include/drivers/display.h

```diff
@@ -286,6 +289,64 @@ static inline void *display_get_framebuffer(const struct device *dev)
 	return api->get_framebuffer(dev);
 }
 
+/**
+ * @brief Set framebuffer for direct access
+ *
+ * @param dev Pointer to device structure
+ *
+ * @param fb Pointer to framebuffer
+ *
+ * @retval 0 on success else negative errno code.
+ *
+ */
+static inline int display_set_framebuffer(const struct device *dev, void *fb)
```

Comment:
IMHO we have to specify in very detail how this API will should work.

1. How to restore original frame buffer (the one accessed by write API before this call) ?
2. Will write API modify the contents of the provided memory?
3. What will happen if we set different pixel format (especially one requiring larger frame buffer)?
    (are we going to ignore potential out-of-bound access by write API? restore default frame buffer?)
4. Should I call this every time the frame buffer is being updated to display changes?
    Is the DMA access continuous (like in the memory-less LCD) or we are just using DMA to update memory in the LCD? (fallback to write API suggests second option).  If this is one-shot API how I can know that DMA operation ended and I can free the buffer?

Reply:
> IMHO we have to specify in very detail how this API will should work.
[Jun] Yes, I will add more comments into the code.
> 1. How to restore original frame buffer (the one accessed by write API before this call) ?
[Jun] I am not sure if every display driver/HW can support this. Actually, I call "write" API inside display_set_framebuffer in case display controller does not support it.
NXP LCD controller has two display buffers, current buffer and prepare buffer. Current buffer is what HW is using and prepare buffer is for SW to set. After SW set the prepare buffer, the current buffer will point to the prepare buffer when display interrupt is triggered.
The driver callback of display_set_framebuffer set the the HW prepare buffer and save this point in the driver as well.
User is supposed to call "display_get_framebuffer " after calling "display_set_framebuffer" to display the next frame.
"display_get_framebuffer" will be blocked until display interrupt comes that last frame is fully loaded into LCD.
> 2. Will write API modify the contents of the provided memory?
[Jun] Legacy write API copies the data to provided memory by display driver. This write API is not sufficient to display camera stream since it's hard to provide a sync mechanism between video in and video out. PS, the SW copy is not good for performance. 
> 3. What will happen if we set different pixel format (especially one requiring larger frame buffer)?
>    (are we going to ignore potential out-of-bound access by write API? restore default frame buffer?)
[Jun] In my sample code, the camera configuration is associated to "display_get_capabilities" for resolution and pixel format of FB.
> 4. Should I call this every time the frame buffer is being updated to display changes?
>    Is the DMA access continuous (like in the memory-less LCD) or we are just using DMA to update memory in the LCD? (fallback to write API suggests second option).  If this is one-shot API how I can know that DMA operation ended and I can free the buffer?
[Jun] This is what we need to add additional display interface to make sure FB is fully loaded into LCD then we can restore the FB. As I comment above, this is only for the display HW that does not support "display_set_framebuffer" and "display_get_framebuffer".

Reply:
@pizi-nordic 
my comments yesterday maybe is incorrect.
In order to be more friendly, I just changed the display API . PS, the display write API is also working for camera stream display, but there is memcpy to consume some CPU cycles.
The display API I added is only used for zero copy of display as below.
static inline int display_set_framebuffer(const struct device *dev, void *fb,
				void **last_fb);
Where last_fb is the point returned by API to the last frame buffer.
Basic work flow of camera is:

 camera_get_prime/secondary->camera_power->camera_get_cap->camera_configure->camera_start->camera_acquire_fb->display->camera_release_fb(last fb)---
                         ^                                                                                             |
                          |______________________________________________________________|

Reply:
@pizi-nordic 
Remove the display code change from this PR, only keep the camera related.
And split some commits.

Reply:
Thank you @JunYangNXP. Could you please put some API description into camera.h?
I would like to fully understand the model for camera devices you propose.

Reply:
Hi @pizi-nordic 
I add comments for each API in header files.
I moved camera API for low level drivers from camera.h to a new camera_drv.h.
So for applications, it only needs to access camera device by API in camera.h.
For camera drivers, it only needs to access camera subsystem by API in camera_drv.h.

Reply:
Thank you @JunYangNXP. At some (high) level the API looks similar to the one proposed in https://github.com/zephyrproject-rtos/zephyr/pull/17194. Do you think that you and @loicpoulain will be able to work together in order have one set of API satisfying needs of both sides?

Reply:
@pizi-nordic 
I had discussion with @loicpoulain about this already. You could read our comments.
Still I recommend the video subsystem designed by @loicpoulain to be used for SW/HW video codec for which camera could be input device. 
For camera subsystem, according to the request from current IoT device, we could need to support:
1) Support multiple cameras on board to capture images from different scopes, something like smartphone which has back/front cameras.
2) User application should not access images sensor directly. Image sensor should be only accessed by camera driver. User application only needs to care about the camera device.
3) Most interface of image sensor present on board is generic and supports plug/unplug. Software should support this to probe image sensor in run-time and no SW update if image sensor is changed.
PS, the performance of NXP camera under @loicpoulain 's video framework is not as expected.

Reply:
IMHO in @loicpoulain solution there are possibility to have multiple cameras as multiple devices. However I do not know, if such approach satisfy your needs.

I personally would like to have a bit more flexible "Video 4 Zephyr" API taking under account all aspects of the video capture and processing instead of set of separate APIs, each supporting only one class of devices.

I also think, that both #17194 and this PR does not provide solid model of camera/video device (and corresponding processing pipeline) which could be used in scenarios different than the one presented in the attached sample.

I recommend starting from setting the requirements and modelling the high-level API, similarly like we did with the logger: https://github.com/zephyrproject-rtos/zephyr/issues/6186#issuecomment-383161969

Reply:
Hi @JunYangNXP,

> Still I recommend the video subsystem designed by @loicpoulain to be used for SW/HW video codec for which camera could be input device.

There is no need to add complex layering, the proposed generic Video API, which has been reworked according to various comments since your review, is perfectly able to manage camera video devices by its own. Your API version is very similar at the end (I see you abstracted sensor bus, which is the right way to go), so I'm not sure we are close to agree.

>     1. Support multiple cameras on board to capture images from different scopes, something like smartphone which has back/front cameras.

Zephyr device/driver architecture perfectly supports, I don't understand the point here.

>     2. User application should not access images sensor directly. Image sensor should be only accessed by camera driver. User application only needs to care about the camera device.

This is something I reworked in #17194 (since your argument of making it simpler was valid), now 'user' interface with the main video device only. However, note that we don't refer to desktop users here but people who develop embedded projects, hackers, etc... So we don't really want to hide devices/hardware but make them easily accessible via a set of APIs.

>     3. Most interface of image sensor present on board is generic and supports plug/unplug. Software should support this to probe image sensor in run-time and no SW update if image sensor is changed.

Again, I2C devices are 'enumerable' via device-tree at build time, not at runtime, like any other SPI, UART, etc device. If you want to support multiple devices, add the corresponding entries as I2C child nodes, with a valid compatible string and corresponding drivers.

>        PS, the performance of NXP camera under @loicpoulain 's video framework is not as expected.

As I said, it's not due to the 'framework', but It the way camera is configured by the initial basic MT9M114 driver, 640x480@7hz. Not sure what is the expectation,  I believe you refer to 30 fps, 60? I don't think it's a problem for the samples, but I'm really open to any change here improving the support. If you agree I (or you) can merge both MT9M114 versions, let me know?

Reply:
Hi @pizi-nordic,
I think both #17194 and this PR really support processing pipeline. And I also don't get what you said "each supporting only one class of devices.". The implementation of pipeline really depends on specific HW driver design. Subsystem and API only provide interface to user.
Hi @pizi-nordic 
Just see your latest code, please forget the 2):)
About 3), I ever talked with @galak , actually, I2C device supports to be enumerable in run-time, but maybe it's not generically used.
About performance, you could refer my PR to configure camera/sensor to exclude the suspicion of your framework.